Move QgsRasterLayerProperties class  to the gui API

vcloarec opened the pull request:
## Description
Motivation :
Move the `QgsRasterLayerProperties` class to the GUI API.
So, this class could be used using the QGIS API, notably for standalone applications.

The instance of this class was dependent, since QGIS 3.8, from the `QgisApp` instance, when calling the `askUserForDatumTransform()`. As this method, defines a title and only call `QgsDatumTransformDialog::run()` method from the gui API, this PR purpose to :
- move the files of the `QgsRasterLayerProperties` class ant the `QgsMapLayerStyleGuiUtils` class to the gui sources
- make those classes public in  the GUI API (GUI_EXPORT)
- change some `#include` and CMakeList.txt files in accordance



## Checklist
- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

PeterPetrik commented:
what is the status of this PR? shall we merge it rather sooner than later?

PeterPetrik commented:
@nyalldawson are you happy to merge it?

vcloarec commented:
merging will be appreciated!

nyalldawson commented on python/gui/auto_generated/qgsmaplayerstyleguiutils.sip.in:
```diff
@@ -0,0 +1,51 @@
+
+
+
+class QgsMapLayerStyleGuiUtils : QObject
```
Does this need to be exposed to python?

nyalldawson commented on python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in:
```diff
@@ -0,0 +1,59 @@
+%End
+
+
+  public slots:
```
This should be made private -- we don't want to expose this

vcloarec replied:
It seems SIP doesn't like when I make this slot private, I am investigating why... I've got a compilation error from a file related to SIP (sip_guipart0.cpp) that contains the method `sipQgsRasterLayerProperties::optionsStackedWidget_CurrentChanged(int a0)`.
The error message says that `QgsRasterLayerProperties::optionsStackedWidget_CurrentChanged(int)` is private. I am not sure to understand why this method is used elsewhere than in `QgsRasterLayerProperties`...

vcloarec replied:
Not expose it for python or for python and c++ API ? When I make this slot private (c++), I've got a compile error from a file related to sip even if run the sipify_all.sh script... I have to investigate why but I have difficulties to understand this sip issue...

nyalldawson commented on python/gui/auto_generated/raster/qgsrasterlayerproperties.sip.in:
```diff
@@ -0,0 +1,59 @@
+%End
+  public:
+
+    enum StyleType
```
Hide this from the python bindings too, by wrapping the definition in the .h file with

```
#ifndef SIPRUN

...
#endif
```

nyalldawson commented on src/gui/raster/qgsrasterlayerproperties.cpp:
```diff
@@ -1272,7 +1273,7 @@ void QgsRasterLayerProperties::pbnAddValuesManually_clicked()
 void QgsRasterLayerProperties::mCrsSelector_crsChanged( const QgsCoordinateReferenceSystem &crs )
 {
-  QgisApp::instance()->askUserForDatumTransform( crs, QgsProject::instance()->crs(), mRasterLayer );
+  QgsDatumTransformDialog::run( crs, QgsProject::instance()->crs(), this, mMapCanvas, tr( "Select Transformation for the raster layer" ) );
```
```suggestion
  QgsDatumTransformDialog::run( crs, QgsProject::instance()->crs(), this, mMapCanvas, tr( "Select Transformation" ) );
```

nyalldawson commented:
You'll need to update the sip bindings after making it private - run scripts/sipify_all.sh

vcloarec commented:
> run scripts/sipify_all.sh

Sorry, I didn't have time to look at this. I've tried to "sipify_all", but compile error still here.
If I make it protected, no problem. But I suppose it is not satisfying...

vcloarec committed:
Move QgsRasterLayerProperties class  to the gui API

Motivation :
Move the QgsRasterLayerProperties class to the GUI API.
So, this class could be used using the QGIS API, notably for standalone applications.

The instance of this class was dependent, since QGIS 3.8, from the QgisApp instance, when calling the askUserForDatumTransform(). As this method, defines a title and only call QgsDatumTransformDialog::run() method from the gui API, this PR purpose to :
- move the files of the QgsRasterLayerProperties class ant the QgsMapLayerStyleGuiUtils class to the gui sources
- make those classes public in  the GUI API (GUI_EXPORT)
- change some #include and CMakeList.txt files in accordance
Makes private and documentation

Makes private some methods of the newcomer class in the gui API and fills documentation
Fix SIP and code layout issues




.
.
conflict resolved
.

vcloarec commented:
@nyalldawson 
the enum isn't exposed to SIP anymore. The slot method is now protected and not expose to SIP. If this method is private, QGIS doesn't compile because a SIP issue.

vcloarec committed:
changes for SIP


.
.


.

vcloarec commented:
ok, as a newbie with SIP, I was quite long to understand all the whys and the wherefores.

Now, only the constructor of the `QgsRasterLayerProperties` class is exposed to python. Maybe it could be useful to open this widget from python code.

If, this PR seems good to you, merging will be very appreciate, and I am open for discussion.

nyalldawson commented:
Perfect -- thanks for seeing this one through!
